sockeye/data_io.py

```diff
@@ -1477,10 +1546,109 @@ def load_state(self, fname: str):
         pass
 
 
+class BatchedRawParallelSampleIter(BaseParallelSampleIter):
```

Comment:
Just an idea: the signature of the constructor of this class is not necessarily set in stone. Instead of creating the SequenceReaders for source/target outside, you could re-create them for every epoch in the implementation of this class, whenever `reset()` is called. Essentially, keep the sequence reader states (which are themselves streaming from disk) as states in the class below.